storage/database/db_manager.go

```diff
@@ -1436,6 +1441,110 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
+
+type badBlock struct {
+	Header *types.Header
+	Body   *types.Body
+}
+
+// badBlockList implements the sort interface to allow sorting a list of
+// bad blocks by their number in the reverse order.
+type badBlockList []*badBlock
+
+func (s badBlockList) Len() int { return len(s) }
+func (s badBlockList) Less(i, j int) bool {
+	return s[i].Header.Number.Uint64() < s[j].Header.Number.Uint64()
+}
+func (s badBlockList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+// ReadBadBlock retrieves the bad block with the corresponding block hash.
+func (dbm *databaseManager) ReadBadBlock(hash common.Hash) *types.Block {
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil
+	}
+	var badBlocks badBlockList
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil
+	}
+	for _, bad := range badBlocks {
+		if bad.Header.Hash() == hash {
+			return types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions)
+		}
+	}
+	return nil
+}
+
+// ReadAllBadBlocks retrieves all the bad blocks in the database.
+// All returned blocks are sorted in reverse order by number.
+func (dbm *databaseManager) ReadAllBadBlocks() ([]*types.Block, error) {
+	var badBlocks badBlockList
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil, err
+	}
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil, err
+	}
+	blocks := make([]*types.Block, 0)
+	for _, bad := range badBlocks {
+		blocks = append(blocks, types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions))
+	}
+	return blocks, nil
+}
+
+// WriteBadBlock serializes the bad block into the database. If the cumulated
+// bad blocks exceeds the limitation, the oldest will be dropped.
+func (dbm *databaseManager) WriteBadBlock(block *types.Block) {
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		logger.Warn("Failed to load old bad blocks", "error", err)
+	}
+	var badBlocks badBlockList
+	if len(blob) > 0 {
+		if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+			logger.Error("failed to decode old bad blocks")
+		}
+	}
+
+	for _, badblock := range badBlocks {
+		if badblock.Header.Hash() == block.Hash() && badblock.Header.Number.Uint64() == block.NumberU64() {
+			logger.Info("There is already corresponding badblock in db.", "badblock number", block.NumberU64())
+			return
+		}
+	}
+
+	badBlocks = append(badBlocks, &badBlock{
+		Header: block.Header(),
+		Body:   block.Body(),
+	})
+	sort.Sort(sort.Reverse(badBlocks))
+	if len(badBlocks) > badBlockToKeep {
+		badBlocks = badBlocks[:badBlockToKeep]
+	}
+	data, err := rlp.EncodeToBytes(badBlocks)
+	if err != nil {
+		logger.Crit("Failed to encode bad blocks", "err", err)
+		return
+	}
+	if err := db.Put(badBlockKey, data); err != nil {
+		logger.Crit("Failed to write bad blocks", "err", err)
+		return
+	}
+}
+
+// DeleteBadBlocks deletes all the bad blocks from the database. Not used anywhere
```

Comment:
How about relocating this definition to the test file?

Reply:
Umm, why? This comment is about definition of DeleteBadBlocks.

Reply:
I meant the definition is a function definition. I think it seems better to relocate this definition to test file since this function is for a test purpose.

Reply:
@hqjang-pepper Is it likely to be used in future work? If so, what about putting TODO? If it is only used at test file only, i also think moving to the test file of deleting is better as @hyunsooda said.